lib/cartopy/mpl/gridliner.py

```diff
@@ -1239,6 +1233,26 @@ def _axes_domain(self, nx=None, ny=None):
 
         return lon_range, lat_range
 
+    def get_visible_children(self):
+        r"""Return a list of the visible child `.Artist`\s."""
+        all_children = (self.xline_artists + self.yline_artists
+                        + self.label_artists)
+        return [c for c in all_children if c.get_visible()]
+
+    def get_window_extent(self, renderer=None):
+        self._draw_gridliner(renderer=renderer)
+        visible_text_extents = [c.get_window_extent(renderer=renderer)
+                                for c in self.label_artists if c.get_visible()]
```

Comment:
Although this works for all tested cases, it feels a bit wrong to be excluding the lines from the bbox calculation here.  Would it be better to define `get_tightbbox` directly as a union of the tightbboxes of the child artists?

Reply:
I think that makes sense to get the union of all child artists as well because they are a part of this artist.

One of the other things I noticed with this when testing it out is if you pan up, the title moves locations based on the axis labels box on the y-axis, which doesn't happen in plain MPL. If the text for the axis extends beyond the spine, then the title gets pushed up with it. Then when the label disappears the title jumps back down again. So I'm wondering if this also needs some adjustment on that aspect as well... (example: https://github.com/SciTools/cartopy/issues/2246)

Reply:
I can reproduce the oscillating title with `main` and with v0.22.  The culprit is this method:

https://github.com/SciTools/cartopy/blob/f2bb81d6bc3826a73ff4e0f67a28c0b04f3ab179/lib/cartopy/mpl/geoaxes.py#L533-L566

Reply:
I added a commit that addresses the case above: we check whether the horizontal extents of the left/right labels overlap the horizontal extent of the axes.  If not, the labels do not affect the title position.  I'm not sure whether this is too simplistic or too projection-specific though.

We do not seem to have any existing test coverage for this title positioning method.  So I tried this example from when the code went in: https://github.com/SciTools/cartopy/pull/1117#issuecomment-445866100  The good news is that I get identical results with my branch and with `main`.  The bad news is I don't get what was shown in that comment.  Note the titles for `AlbersEqualArea` and `LambertConformal` are lower here than they were then.  A closer look at the `AlbersEqualArea` example shows that it has no "left", "right" or "top" labels, only "bottom" and "geo".  So it seems like "geo" needs to be incorporated into the method somehow.  I haven't really got my head around what the "geo" labels do though.

![gridliner_titling global branch](https://github.com/SciTools/cartopy/assets/10599679/407682fb-7373-498a-8c00-6805eabeffa2)

Reply:
OK, I think I get it now: the left and right labels refer to anything on the vertical spines of the standard rectangular axes.  The "geo" labels are everywhere where the spine curves round (I think).  That example now gives me the below image, and I have pulled some of it out into a new image test.

![gridliner_titling global](https://github.com/SciTools/cartopy/assets/10599679/7bdf2735-e64a-4f51-a57e-22a7afec71a0)